### PR TITLE
Fix Issue 2765 - rule option: migrate to GeoIP2 database and libmaxminddb API.

### PR DESCRIPTION
**Pull request by bmeeks8**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x ] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [ x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x ] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: [#2765](https://redmine.openinfosecfoundation.org/issues/2765)

Describe changes:
- Update "geoip" rule option support code to use the libmaxminddb library for reading "iso_country" code for an IP address from the GeoIP2/GeoLite2 database format. The legacy GeoIP database format and supporting library (libgeoip) has been discontinued by MaxMind and replaced by GeoIP2 and the libmaxminddb library.

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

### CODE REVIEW

**Comment by bmeeks8** [comment]
Apparently there is no _libmaxminddb_ library package available for Cygwin, thus the test build on appveyor is failing because it says it cannot find the library.  I've searched the web and did not find a binary package for Cygwin I could reference.  I did find source code for maybe creating a binary package, though, using _cygport_.

So does this mean "geoip" support in Suricata is going to die with the end-of-life of the legacy GeoIP database, or is there another alternative for including the GeoIP2 format supported by _libmaxminddb_ in Suricata?

File: src/detect-geoip.c [hunk 1]
```diff
@@ -84,27 +84,85 @@ void DetectGeoipRegister(void)
...
-static GeoIP *InitGeolocationEngine(void)
+static int InitGeolocationEngine(DetectGeoipData *geoipdata)
 {
-    return GeoIP_new(GEOIP_MEMORY_CACHE);
+    const char *filename = NULL;
+
+    /* Get location and name of GeoIP2 database from YAML conf */
+    (void)ConfGet("geoip-database", &filename);
+
+    if (filename == NULL) {
```
Comment (victorjulien) [inline src/detect-geoip.c]: would it make sense to use a default location in this case?

File: src/detect-geoip.c [hunk 2]
```diff
@@ -84,27 +84,85 @@ void DetectGeoipRegister(void)
...
+    if (result.found_entry) {
+        mmdb_error = MMDB_get_value(&result.entry, &entry_data, "country", 
+                                    "iso_code", NULL);
+        if (mmdb_error != MMDB_SUCCESS)
+            return NULL;
+
+        /* If ISO country code was found, then return it */
+        if (entry_data.has_data) {
+            if (entry_data.type == MMDB_DATA_TYPE_UTF8_STRING) {
+                    char *country_code = strndup((char *)entry_data.utf8_string,
```
Comment (victorjulien) [inline src/detect-geoip.c]: Why is the string copied here? It seems unnecessary. If it is needed, please use SCStrdup instead.

File: src/detect-geoip.c [hunk 3]
```diff
@@ -125,29 +183,44 @@ static const char *GeolocateIPv4(GeoIP *geoengine, uint32_t ip)
...
+        return 0;
+
     /* Check if NOT NEGATED match-on condition */
     if ((geoipdata->flags & GEOIP_MATCH_NEGATED) == 0)
     {
-        for (i = 0; i < geoipdata->nlocations; i++)
-            if (country != NULL && strcmp(country, (char *)geoipdata->location[i])==0)
+        for (i = 0; i < geoipdata->nlocations; i++) {
+            if (country != NULL && strcmp(country, (char *)geoipdata->location[i])==0) {
+                free(country);
```
Comment (victorjulien) [inline src/detect-geoip.c]: SCFree

**Comment by victorjulien** [comment]
Our Windows support efforts are now focused on MinGW, which does have a libmaxminddb package. In any case I would like to get this work merged, even if not all platforms support it.

File: src/detect-geoip.c (hunk 3 above)
Comment (bmeeks8) [reply src/detect-geoip.c]: Will change this.

File: src/detect-geoip.c (hunk 1 above)
Comment (bmeeks8) [reply src/detect-geoip.c]: The new MaxMind library, when passed a NULL filename, will return a file open error and terminate.  This behavior is different from the legacy libGeoIP library which would search a default path for a DB file before giving up.

Do you want to include all the code baggage of locating a default path and DB file across Linux/Windows platforms, or just log the error message if a valid DB file and path is not given in suricata.yaml?

Comment (bmeeks8) [reply src/detect-geoip.c]: After further reflection, I propose leaving the code "as-is" whereby it logs an appropriate warning message and disables the geoip rule option when the GeoIP2 database is not specified in _suricata.yaml_.  The _libmaxminddb_ library will not use a default path when NULL is provided for the DB name.  There are at least two different DB names possible for the Country database depending on whether the user has the free GeoLite2 edition or the paid subscription GeoIP2 edition.  The user is responsible for configuring the DB download anyway, so the user may as well specify the location of the DB to Suricata via the _suricata.yaml_ entry.

**Comment by bmeeks8** [comment]
I believe all review comments have been resolved and this request is ready for acceptance and merging.

File: src/detect-geoip.c (hunk 2 above)
Comment (victorjulien) [reply src/detect-geoip.c]: Ok fair enough, but then I would prefer a proper SCStrndup implementation to be introduced. It should be quite simple to add that.

**Pull request closed without merging**